### ISSUE TITLE
Corrupted mail body under Node v10.11.0

### ISSUE THREAD

**Issue by kolontsov** [haraka/Haraka#2526]
### system info

Haraka | Haraka.js — Version: 2.8.21
 --- | :--- 
Node | v10.11.0
OS | Linux hostname 3.13.0-143-generic #192-Ubuntu SMP Tue Feb 27 10:45:36 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
openssl | OpenSSL 1.0.1f 6 Jan 2014

### Expected behavior

Mail body is never corrupted.

### Observed behavior

Sometimes first line of the body is prepended by part of random header. For example:

```
[..more headers..]
Content-Type: text/plain; charset=utf-8; format=flowed
DKIM-Signature: v=1;a=rsa-sha256;bh=nyCg05uC/17m0kK8XW7GYE/Cqw/9CGnGx/LqRxCEsI8=;c=relaxed/simple;d=somedomain.com;h=from;s=dkim;b=Ckip0bjX6DCA88gyK7dEid8ENml8Z1CZOk+/mE8lkFgatxbzQ3STERLDesKIc4vF0mCV8cjbwdCoTXS7qZfR6g5dEjWImAMMoYlkH8kirRSrVamhIuFGSISkP8ZdFuOxasa3N08uhUU/es5c8N3YP6jo6zb7OYJKcPhc8oInqxg=

Content-Type: text/plain; charset=utHi,    <-- body starts here

I hope all is well [...more body text..]
```

### Steps to reproduce

Install Node v10+, set `spool_after=0` in `smtp.ini`, wait for corrupted mail, it will come soon. Tested under Ubuntu 14.04 and 16.04. Also see demo scripts below.

The issue: when you read from file stream, it's not guaranteed that file, or at least it's "header part" will be fully contained in first chunk - even if it's really small (for example, <1kb) and highWaterMark is 64k. And it's especially true under Node v10, where you can easily get chunk with less than few hundred bytes in size if you have a lot of concurrent streams. 

The problem is here: https://github.com/haraka/Haraka/blob/e0598e737cb1286e2096ab798bc7b4167befc2d8/messagestream.js#L263-L269

We're skipping headers here. But if not all headers are inside the chunk, because chunk is short and "truncated" (for example, end of buffer is `\r\nContent-Type: text/plain; charset=ut`), then the loop will exit after last CR/LF, and this incomplete part will be added to the output because:

https://github.com/haraka/Haraka/blob/e0598e737cb1286e2096ab798bc7b4167befc2d8/messagestream.js#L295-L299

(by the way, another possible issue is CR/LF splitted between different chunks)

Demo with messagestream (adjust ATTEMPTS/DELAY if needed):
```javascript
const stream = require('stream');
const fs = require('fs');
const assert = require('assert');
const MessageStream = require('./messagestream');
const ATTEMPTS = 2048;
const DELAY    = 3000;

let headers = [];
for (let i=0; i<48; i++)
    headers.push(`Another-Header-Name${i}: abcdefghijklmnopqrstuvwxyz\r\n`);

let test = id=>{
    let msg = new MessageStream({main: {spool_after: 0}}, id, headers);
    try { fs.unlinkSync(msg.filename); } catch(e){};
    headers.forEach(h=>msg.add_line(h));
    msg.add_line('\r\n');
    msg.add_line('Hello\r\n');
    msg.add_line_end();
    msg.get_data(data=>{
        let eml = data.toString().split(/\r\n/);
        if (!eml[eml.length-2].startsWith('Hello'))
        {
            console.log('\nfname = '+msg.filename);
            for (let i=eml.length-5; i<eml.length; i++)
                console.log(eml[i]);
            process.exit(0);
        }
    });
};

for (let i=1; i<ATTEMPTS; i++)
    setTimeout(()=>test(i), Math.random()*DELAY);
```

My output: 
```
fname = /tmp/218.eml
Another-Header-Name46: abcdefghijklmnopqrstuvwxyz
Another-Header-Name47: abcdefghijklmnopqrstuvwxyz

Another-Header-Name37: abcdefghijklmnHello
```

Another demo which shows how frequently you can get multiple chunks reading single tiny file concurrently:

```javascript
const fs = require('fs');
const ATTEMPTS = 1024;
const DELAY = 3000;
const fname = '/etc/profile';
const fsize = fs.statSync(fname).size;

let cnt = 0, mismatch = 0;

let test = (id, hw)=>{
    let s = fs.createReadStream(fname);
    s.on('data', chunk=>{
        if (chunk.length!=fsize)
        {
            console.log('attempt #'+id+': chunksize '+chunk.length+' ');
            mismatch++;
        }
    });
    s.on('end', chunk=>++cnt==ATTEMPTS && console.log('mismatch: '+mismatch));
};

console.log('Reading '+fname+', size: '+fsize);
for (let i=0; i<ATTEMPTS; i++)
    setTimeout(()=>test(i), Math.random()*DELAY);
```

My output:
```
Reading /etc/profile, size: 684
attempt #215: chunksize 556 
attempt #215: chunksize 128 
attempt #216: chunksize 556 
attempt #216: chunksize 128 
attempt #227: chunksize 556 
attempt #227: chunksize 128 
attempt #592: chunksize 300 
attempt #592: chunksize 384 
attempt #222: chunksize 172 
attempt #222: chunksize 512 
attempt #1011: chunksize 428 
attempt #1011: chunksize 256 
attempt #638: chunksize 300 
attempt #638: chunksize 384 
mismatch: 14
```

**Comment by smfreegard** [comment]
And I just realised - you aren't using `chunkemitter` in your test code (which is what the stream uses), so your results will be different.

**Comment by kolontsov** [comment]
> Why are you setting spool_after=0 instead of leaving it as the default (e.g. don't spool).

I set `spool_after=0` to demonstrate issues with current code reading mails from `spool_dir`. When you keep everything in a memory, it's not a problem. Ability to use `spool_after=-1` doesn't mean that `messagestream.process_buf` doesn't have a bug.

> And I just realised - you aren't using chunkemitter in your test code

As I said, results are 100% reproducible on a real server. I submitted two variants of test code - to demonstrate what exactly goes wrong. The first variant uses `messagestream`, which uses `chunkemitter` internally (but not for reading body from spool_dir..). The second variant is a just a short example to demonstrate why `messagestream` puts wrong data to `read_ce` chunkemitter.

**Comment by kolontsov** [comment]
(just to clarify, you don't need exactly `spool_after=0` to get problems on the server: any value >=0 (even large enough) could result in body corruption of emails which satisfy this condition).

**Comment by algohali3008** [comment]
Thanks

**Comment by sathishvenkataraman** [comment]
Hi, i face the same issue, will this be fixed? any alternatives?

**Comment by kolontsov** [comment]
> Hi, i face the same issue, will this be fixed? any alternatives?

I just decided to use `spool_after=-1`

**Comment by sathishvenkataraman** [comment]
> > Hi, i face the same issue, will this be fixed? any alternatives?
> 
> I just decided to use `spool_after=-1`
> 
> (@msimerson, sorry for not answering in time! I missed it somehow)

Thanks, but i need the spooling feature badly, hence checking if there is a way to do it without the data getting corrupted or a fix available for this.

**Comment by sathishvenkataraman** [comment]
Any help on resolving this issue?

**Comment by msimerson** [comment]
I think this is fixed by #3068
